client/src/main/java/io/pravega/client/stream/impl/SingleRoutingKeyTransactionWriterImpl.java

```diff
@@ -0,0 +1,204 @@
+                    state = Transaction.Status.COMMITTING;
+                } else {
+                    throw new IllegalStateException("Transaction is already closed.");
+                }
+            }
+            return write.apply(events);
+        }
+
+        @Override
+        public void abort() {
+            synchronized (this) {
+                if (state.equals(Transaction.Status.OPEN)) {
+                    state = Transaction.Status.ABORTED;
+                }
+            }
+        }
+
+        private void throwIfClosed() throws TxnFailedException {
+            synchronized (this) {
+                if (!state.equals(Transaction.Status.OPEN)) {
+                    throw new TxnFailedException();
+                }
+            }
+        }
+    }
+
+    @Override
+    public SingleRoutingKeyTransaction<Type> beginTxn() {
+        synchronized (this) {
+            if (transaction == null) {
+                transaction = new SingleRoutingKeyTransactionImpl<>(serializer, this::writeEvents);
+                return transaction;
+            } else {
+                throw new IllegalStateException();
+            }
+        }
+    }
+
+    private SegmentOutputStream getSegmentOutputStream() {
+        synchronized (this) {
+            if (segmentOutputStream != null) {
+                return segmentOutputStream;
+            } else {
+                return controller.getCurrentSegments(stream.getScope(), stream.getStreamName())
+                                 .thenApply(segments -> {
+                                     Segment segment = segments.getSegmentForKey(routingKey);
+                                     return createSegmentOutputStream(segment);
+                                 }).join();
+            }
+        }
+    }
+
+    private SegmentOutputStream createSegmentOutputStream(Segment segment) {
+        synchronized (this) {
+            if (!segmentOutputStream.getSegmentName().equals(segment.getScopedName())) {
+                segmentOutputStream = outputStreamFactory.createOutputStreamForSegment(segment,
+                        this::segmentSealedCallback,
+                        EventWriterConfig.builder().build(), tokenProvider);
+            }
+            return segmentOutputStream;
+        }
+    }
+
+    @Synchronized
+    private void segmentSealedCallback(Segment segment) {
+        // instead of traversing thru successors we will call current segments and fetch the appropriate segment.
+        // by setting segmentOutputStream to null we will force the caller to get the replacement segmentoutputstream.
+        // we will call resend while holding the lock so that successor segment outputstream is setup and events are
+        // replayed to it before new writeevents into successor are triggered. 
+        List<PendingEvent> toResend = segmentOutputStream.getUnackedEventsOnSeal();
+        segmentOutputStream = null;
+        for (PendingEvent event : toResend) {
```

Comment:
I do not see how there can be multiple PendingEvents here. Am I missing something here?

Reply:
so conceptually there could be multiple "transactions" created on the same txn writer. so there could be more than one pending event (one per transaction)